app/javascript/lib/scp-api.jsx

```diff
@@ -250,8 +250,11 @@ export function setupRenewalForReadOnlyToken(studyAccession) {
  */
 export async function checkTerraTosAcceptance() {
   const apiUrl = '/site/check_terra_tos_acceptance'
-  const [response] = await scpApi(apiUrl, defaultInit())
-  return response.mustAccept
+  await scpApi(apiUrl, defaultInit()).then(response => {
+    return { mustAccept: response.mustAccept, status: response.status }
+  }).catch(error => {
+    return { mustAccept: error.mustAccept, status: error.status }
+  })
```

Comment:
Blocker: this breaks when the user must accept updated Terra terms of service.

I just manually tested this locally, and `checkTerraTosAcceptance` now returns `undefined`.  Here's the downstream code that calls this, [in application.js](https://github.com/broadinstitute/single_cell_portal_core/blob/f7a1eba618fcfd35a114dd7b83be410dc679b3d4/app/javascript/vite/application.js#L57):

```
    ScpApi.checkTerraTosAcceptance().then(mustAcceptTerraTos => {
      if (mustAcceptTerraTos) {
        window.location = '/single_cell/terra_tos'
      }
    })
```

So with `mustAcceptTerraTos` always returning `undefined`, this new code will always give a negative (often false negative) result.  In other words, it will always effectively behave as if the user does not need to accept Terra ToS.

Technical issues:

1. The `return` is inside a callback function, and doesn't get assigned to anything in the containing scope.  That's why `checkTerraTosAcceptance` always returns `undefined`.

2. [`scpApi` returns an array](https://github.com/broadinstitute/single_cell_portal_core/blob/f7a1eba618fcfd35a114dd7b83be410dc679b3d4/app/javascript/lib/scp-api.jsx#L912) when the response is OK.

3. `scpApi` indeed throws an error upon 401 or 403.  But lines 932-948 in scp-api.jsx entail that if the response status is 401 or 403, then the return value of `scpApi` will be a string, not a JSON object with a `mustAccept` or `status` property.  The UI throws an error modal saying "Authentication failed Your session may have timed out. Please sign in again."  Now this API endpoint returns a more specific 401 or 403, that modal is misleading, and in this case we should ensure the frontend executes `window.location = '/single_cell/terra_tos'`. This is the bug.

A full-fidelity test of 401 / 403 handling is impractical, as you note.  However, you can easily and adequately test UI handling by setting [line 81 in `app/controllers/api/v1/site_controller.rb`](https://github.com/broadinstitute/single_cell_portal_core/pull/1787/files#diff-ef6cc11ca821ecf554c8d9a3c345231cf35421b2a16f30449d74eaa4a112b58fR81) to `render json: { must_accept: true }, status: 401`.

Here's an initial pass for part of a fix:

```suggestion
  return await scpApi(apiUrl, defaultInit()).then(([response]) => {
    return response.mustAccept
  }).catch(error => {
    return error.mustAccept
  })
```

Note that the `error` case would require further updates to the 401/403 handling in `scp-api.jsx`.

Reply:
Ah yes, I see what you're saying now.  OK - I think this is still tractable but will require more work.

Reply:
I just realized I'd had this working previously but I took it out in a later commit because I thought it was unnecessary given some other refinements.  I think I have a good idea on how to handle this now - thanks for catching it!